src/test/test_tls.c

```diff
@@ -0,0 +1,507 @@
+        0x00, 0x0f, // Extension Type: Heart Beat
+        0x00, 0x01, // Length
+        0x01 // Mode: Peer allows to send requests
+};
+static struct test_packet good[] = {
+        { (char *)good_data_1, sizeof(good_data_1) },
+        { (char *)good_data_2, sizeof(good_data_2) },
+        { (char *)good_data_3, sizeof(good_data_3) },
+        { (char *)good_data_4, sizeof(good_data_4) },
+        { (char *)good_data_5, sizeof(good_data_5) }
+};
+
+static struct test_packet bad[] = {
+        { (char *)ssl30_request, sizeof(ssl30_request) },
+        { (char *)ssl20_client_hello, sizeof(ssl20_client_hello) },
+        { (char *)bad_data_1, sizeof(bad_data_1) },
+        { (char *)bad_data_2, sizeof(bad_data_2) },
+        { (char *)bad_data_3, sizeof(bad_data_3) }
+};
+
+int main() {
+    uint8_t *pkt = (uint8_t *)good_data_1;
+    int error = 0;
+    char sn[FQDN_LENGTH];
+    memset(sn, 0, FQDN_LENGTH);
+    *sn = 0;
+    error = get_server_name(pkt, sizeof(good_data_1), pkt, sn);
+    assert(strcmp("localhost", sn) == 0);
+    assert(error == strlen(sn));
+
+    pkt = (uint8_t *)good_data_2;
+    memset(sn, 0, FQDN_LENGTH);
+    *sn = 0;
+    error = get_server_name(pkt, sizeof(good_data_2), pkt, sn);
+    assert(strcmp("ddg.com", sn) == 0);
+    assert(error == strlen(sn));
+
+    pkt = (uint8_t *)good_data_3;
+    memset(sn, 0, FQDN_LENGTH);
+    *sn = 0;
+    error = get_server_name(pkt, sizeof(good_data_3), pkt, sn);
+    assert(strcmp("localhost", sn) == 0);
+    assert(error == strlen(sn));
+
+    pkt = (uint8_t *)good_data_4;
+    memset(sn, 0, FQDN_LENGTH);
+    *sn = 0;
+    error = get_server_name(pkt, sizeof(good_data_4), pkt, sn);
+    assert(strcmp("localhost", sn) == 0);
+    assert(error == strlen(sn));
+
+    pkt = (uint8_t *)good_data_5;
+    memset(sn, 0, FQDN_LENGTH);
+    *sn = 0;
+    error = get_server_name(pkt, sizeof(good_data_5), pkt, sn);
+    assert(strcmp("localhost", sn) == 0);
+    assert(error == strlen(sn));
+
+    pkt = (uint8_t *)ssl30_request;
+    memset(sn, 0, FQDN_LENGTH);
+    *sn = 0;
+    error = get_server_name(pkt, sizeof(ssl30_request), pkt, sn);
+    assert(strcmp("localhost", sn) != 0);
+    assert(strlen(sn) == 0);
+    assert(error == -2);
+
+    pkt = (uint8_t *)ssl20_client_hello;
+    memset(sn, 0, FQDN_LENGTH);
+    *sn = 0;
+    error = get_server_name(pkt, sizeof(ssl20_client_hello), pkt, sn);
+    assert(strcmp("localhost", sn) != 0);
+    assert(strlen(sn) == 0);
+    assert(error == -2);
+
+    pkt = (uint8_t *)bad_data_1;
+    memset(sn, 0, FQDN_LENGTH);
+    *sn = 0;
+    error = get_server_name(pkt, sizeof(bad_data_1), pkt, sn);
+    assert(strcmp("localhost", sn) != 0);
+    assert(strlen(sn) == 0);
+    assert(error == -4);
+
+    pkt = (uint8_t *)bad_data_2;
+    memset(sn, 0, FQDN_LENGTH);
+    *sn = 0;
+    error = get_server_name(pkt, sizeof(bad_data_2), pkt, sn);
+    assert(strcmp("localhost", sn) != 0);
+    assert(strlen(sn) == 0);
+    assert(error == -1);
```

Comment:
This test case needs an update after we decided not to return when TLS data length is smaller than expected.